synapse/replication/tcp/redis.py

```diff
@@ -0,0 +1,161 @@
+            "replication-" + cmd.get_logcontext_id(), self.handle_command, cmd
+        )
+
+    async def handle_command(self, cmd: Command):
+        """Handle a command we have received over the replication stream.
+
+        By default delegates to on_<COMMAND>, which should return an awaitable.
+
+        Args:
+            cmd: received command
+        """
+        handled = False
+
+        # First call any command handlers on this instance. These are for redis
+        # specific handling.
+        cmd_func = getattr(self, "on_%s" % (cmd.NAME,), None)
+        if cmd_func:
+            await cmd_func(cmd)
+            handled = True
+
+        # Then call out to the handler.
+        cmd_func = getattr(self.handler, "on_%s" % (cmd.NAME,), None)
+        if cmd_func:
+            await cmd_func(cmd)
+            handled = True
+
+        if not handled:
+            logger.warning("Unhandled command: %r", cmd)
+
+    def connectionLost(self, reason):
+        logger.info("Lost connection to redis instance")
+        self.handler.lost_connection(self)
+
+    def send_command(self, cmd):
+        """Send a command if connection has been established.
+
+        Args:
+            cmd (Command)
+        """
+        string = "%s %s" % (cmd.NAME, cmd.to_line())
+        if "\n" in string:
+            raise Exception("Unexpected newline in command: %r", string)
+
+        encoded_string = string.encode("utf-8")
+
+        async def _send():
+            with PreserveLoggingContext():
+                await self.redis_connection.publish(self.stream_name, encoded_string)
+
+        run_as_background_process("send-cmd", _send)
+
+    def stream_update(self, stream_name, token, data):
+        """Called when a new update is available to stream to clients.
+
+        We need to check if the client is interested in the stream or not
+        """
+        self.send_command(RdataCommand(stream_name, token, data))
+
+
+class RedisFactory(txredisapi.SubscriberFactory):
+    """This is a reconnecting factory that connects to redis and immediately
+    subscribes to a stream.
+    """
+
+    maxDelay = 5
+    continueTrying = True
+    protocol = RedisSubscriber
+
+    def __init__(self, hs):
+        super(RedisFactory, self).__init__()
+
+        self.password = hs.config.redis.redis_password
+
+        self.handler = hs.get_tcp_replication()
+        self.stream_name = hs.hostname
+
+        self.redis_connection = txredisapi.lazyConnection(
+            host=hs.config.redis_host,
+            port=hs.config.redis_port,
+            dbid=hs.config.redis_dbid,
+            password=hs.config.redis.redis_password,
+            reconnect=True,
+        )
+
+        self.conn_id = random_string(5)
+
+    def buildProtocol(self, addr):
+        p = super(RedisFactory, self).buildProtocol(addr)
+        p.handler = self.handler
+        p.redis_connection = self.redis_connection
+        p.conn_id = self.conn_id
+        p.stream_name = self.stream_name
```

Comment:
can these be declared in `RedisSubscriber`'s constructor?

Reply:
The library is a bit of a mess, really. If we add these to the protocol constructor then we'll need to fully override `buildProtocol`, which does things other than just instantiate the protocol. So its not really clear how you're meant to use this without doing this or copy and pasting bits of the code around (though the library is designed to be copied into your repo rather than using pip, so maybe that is how you're meant to do it)

Reply:
I've added a comment to that effect for now.

Reply:
> though the library is designed to be copied into your repo rather than using pip

is it? god, that doesn't inspire much confidence. Maybe we should do that then?

TBH my comment here was mostly that I wanted *some* recognition of the existence of these attributes in `RedisSubscriber` - even setting them to `None` is fine, so I think you've addressed this comment.

one thought for a workaround would be to make `self.protocol` a callable which can add a few extra kwargs which `RedisSubscriber.__init__` could intercept. I'm not sure that buys a great deal though.

Reply:
> > though the library is designed to be copied into your repo rather than using pip
> 
> is it? god, that doesn't inspire much confidence. Maybe we should do that then?

Maybe, my bias is against doing so though.